espnet2/tts/gst/style_encoder.py

```diff
@@ -0,0 +1,241 @@
+                torch.nn.ReLU(inplace=True),
+            ]
+        self.convs = torch.nn.Sequential(*convs)
+
+        self.conv_layers = conv_layers
+        self.kernel_size = conv_kernel_size
+        self.stride = conv_stride
+        self.padding = padding
+
+        # get the number of GRU input units
+        gru_in_units = idim
+        for i in range(conv_layers):
+            gru_in_units = (
+                gru_in_units - conv_kernel_size + 2 * padding
+            ) // conv_stride + 1
+        gru_in_units *= conv_out_chans
+        self.gru = torch.nn.GRU(gru_in_units, gru_units, gru_layers, batch_first=True)
+
+    def forward(self, speech: torch.Tensor) -> torch.Tensor:
+        """Calculate forward propagation.
+
+        Args:
+            speech (Tensor): Batch of padded target features (B, Lmax, idim).
+
+        Returns:
+            Tensor: Reference embedding (B, gru_units)
+
+        """
+        batch_size = speech.size(0)
+        xs = speech.unsqueeze(1)  # (B, 1, Lmax, idim)
+        hs = self.convs(xs).transpose(1, 2)  # (B, Lmax', conv_out_chans, idim')
+        # NOTE(kan-bayashi): We need to care the length?
+        time_length = hs.size(1)
+        hs = (
+            hs.transpose(1, 2).contiguous().view(batch_size, time_length, -1)
+        )  # (B, Lmax', gru_units)
+        self.gru.flatten_parameters()
+        _, ref_embs = self.gru(hs)  # (gru_layers, batch_size, gru_units)
+        ref_embs = ref_embs[-1]  # (batch_size, gru_units)
+
+        return ref_embs
+
+
+class StyleTokenLayer(torch.nn.Module):
+    """Style token layer module.
+
+    This module is style token layer introduced in `Style Tokens: Unsupervised Style
+    Modeling, Control and Transfer in End-to-End Speech Synthesis`.
+
+    .. _`Style Tokens: Unsupervised Style Modeling, Control and Transfer in End-to-End
+        Speech Synthesis`: https://arxiv.org/abs/1803.09017
+
+    Args:
+        ref_embed_dim (int, optional): Dimension of the input reference embedding.
+        gst_tokens (int, optional): The number of GST embeddings.
+        gst_token_dim (int, optional): Dimension of each GST embedding.
+        gst_heads (int, optional): The number of heads in GST multihead attention.
+        dropout_rate (float, optional): Dropout rate in multi-head attention.
+
+    """
+
+    def __init__(
+        self,
+        ref_embed_dim: int = 128,
+        gst_tokens: int = 10,
+        gst_token_dim: int = 256,
+        gst_heads: int = 4,
+        dropout_rate: float = 0.0,
+    ):
+        """Initilize style token layer module."""
+        assert check_argument_types()
+        super(StyleTokenLayer, self).__init__()
+
+        gst_embs = torch.randn(gst_tokens, gst_token_dim)
+        self.register_parameter("gst_embs", torch.nn.Parameter(gst_embs))
+        self.projection = torch.nn.Linear(ref_embed_dim, gst_token_dim)
+        self.mha = MultiHeadedAttention(gst_heads, gst_token_dim, dropout_rate)
```

Comment:
I am not fully sure about the dimension of the style token embedding (256 vs. 256//num_heads). According to the paper, 
> When using h attention heads, we set the token embedding size to be 256/h and concatenate the attention outputs, such that the final style embedding size remains the same.

I think the current code would work okay, but if we strictly follow the original paper's description, would it be better to use 256//num_heads dim for the style token embedding? I am wondering like modifying https://github.com/espnet/espnet/blob/6f6452a3891d11b0b0306084dd46a464ddf8e25c/espnet/nets/pytorch_backend/transformer/attention.py#L33-L34 to 
```py
        self.linear_k = nn.Linear(key_dim, n_feat) # 256//num_heads  -> 256
        self.linear_v = nn.Linear(value_dim, n_feat) # 256//num_heads -> 256
```
for example (BUT I am not sure to be honest...) Would be nice to check the TTS quality.

Reply:
Yeah, maybe you are right.
Since our multihead attention assumes the same dimension, I use embedding size = 256 (256 -> 64*4 -> 256).
Let us check the performance after fixing transpose bug ...

Reply:
I added note about the difference.